tsl/src/compression/create.c

```diff
@@ -741,6 +743,207 @@ validate_hypertable_for_compression(Hypertable *ht)
+					 " (SELECT array_agg(x)  "
+					 " FROM jsonb_array_elements_text(seg_by->'columns') t(x))::text[], "
+					 " seg_by->>'message', "
+					 " (seg_by->>'confidence')::int "
+					 "FROM %s.%s(%d) seg_by",
+					 quote_identifier(get_namespace_name(get_func_namespace(default_segmentby_fn))),
+					 quote_identifier(get_func_name(default_segmentby_fn)),
+					 ht->main_table_relid);
+
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "could not connect to SPI");
+
+	res = SPI_execute(command->data, true /* read_only */, 0 /*count*/);
+
+	if (res < 0)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 (errmsg("could not get the default segment by for a hypertable \"%s\"",
+						 get_rel_name(ht->main_table_relid)))));
+
+	old = MemoryContextSwitchTo(upper);
+	datum = SPI_getbinval(SPI_tuptable->vals[0], SPI_tuptable->tupdesc, 1, &isnull);
+	if (!isnull)
+		column_res = DatumGetArrayTypePCopy(datum);
+	MemoryContextSwitchTo(old);
+
+	datum = SPI_getbinval(SPI_tuptable->vals[0], SPI_tuptable->tupdesc, 2, &isnull);
+
+	if (!isnull)
+	{
+		message = DatumGetTextPP(datum);
+		elog(WARNING,
+			 "there was some uncertainty picking the default segment by for the hypertable: %s",
+			 text_to_cstring(message));
+	}
+
+	datum = SPI_getbinval(SPI_tuptable->vals[0], SPI_tuptable->tupdesc, 3, &isnull);
+	if (!isnull)
+	{
+		confidence = DatumGetInt32(datum);
+	}
+
+	res = SPI_finish();
+	if (res != SPI_OK_FINISH)
+		elog(ERROR, "SPI_finish failed: %s", SPI_result_code_string(res));
+
+	result = makeStringInfo();
+	ts_array_append_stringinfo(column_res, result);
+	elog(NOTICE,
+		 "default segment by for hypertable \"%s\" is set to \"%s\"",
+		 get_rel_name(ht->main_table_relid),
+		 result->data);
+
+	elog(DEBUG1,
+		 "segment_by default: hypertable=\"%s\" columns=\"%s\" function: \"%s.%s\" confidence=%d",
+		 get_rel_name(ht->main_table_relid),
+		 result->data,
+		 get_namespace_name(get_func_namespace(default_segmentby_fn)),
+		 get_func_name(default_segmentby_fn),
+		 confidence);
+	return column_res;
+}
+
+/*
+ * Get the default segment by value for a hypertable
+ */
+static OrderBySettings
+compression_setting_orderby_get_default(Hypertable *ht, ArrayType *segmentby)
+{
+	StringInfo command;
+	int res;
+	text *column_res = NULL;
+	Datum datum;
+	text *message;
+	bool isnull;
+	MemoryContext upper = CurrentMemoryContext;
+	MemoryContext old;
+	char *orderby;
+	int32 confidence = -1;
+
+	Oid types[] = { TEXTARRAYOID };
+	Datum values[] = { PointerGetDatum(segmentby) };
+	char nulls[] = { segmentby == NULL ? 'n' : 'v' };
+	Oid orderby_fn = ts_guc_default_orderby_fn_oid();
+
+	if (!OidIsValid(orderby_fn))
+	{
+		/* fallback to original logic */
+		OrderBySettings obs = (OrderBySettings){ 0 };
+		obs = add_time_to_order_by_if_not_included(obs, segmentby, ht);
+		elog(DEBUG1,
+			 "order_by default: hypertable=\"%s\" function=\"\" confidence=-1",
+			 get_rel_name(ht->main_table_relid));
+		return obs;
+	}
+
+	command = makeStringInfo();
```

Comment:
Should you release this memory after you're done?

Reply:
fixed